Library/Homebrew/test/utils/git_spec.rb

```diff
@@ -0,0 +1,175 @@
+require "utils/git"
+
+describe Git do
+  before(:all) do
+    git = HOMEBREW_SHIMS_PATH/"scm/git"
+
+    @file = "lib/blah.rb"
+    @repo = Pathname.new("repo")
+
+    (@repo/"lib").mkpath
+    system git, "init"
+    FileUtils.touch(@repo/@file)
+
+    File.open(@repo/@file, "w") { |f| f.write("blah") }
+    system git, "add", @repo/@file
+    system git, "commit", "-m", "'File added'"
+    @h1 = `git rev-parse HEAD`
+
+    File.open(@repo/@file, "w") { |f| f.write("brew") }
+    system git, "add", @repo/@file
+    system git, "commit", "-m", "'written to File'"
+    @h2 = `git rev-parse HEAD`
+  end
+
+  let(:hash1) { @h1[0..6] }
+  let(:hash2) { @h2[0..6] }
+
+  after(:all) do
+    FileUtils.rm_rf(@repo)
+  end
+
+  describe "#last_revision_commit_of_file" do
+    it "gives last revision commit when before_commit is nil" do
+      expect(
+        described_class.last_revision_commit_of_file(@repo, @file),
+      ).to eq(hash1)
+    end
+
+    it "gives revision commit based on before_commit when it is not nil" do
+      expect(
+        described_class.last_revision_commit_of_file(@repo,
+                                                    @file,
+                                                    before_commit: hash2),
+      ).to eq(hash2)
+    end
+  end
+
+  describe "#last_revision_of_file" do
+    it "returns last revision of file" do
+      expect(
+        described_class.last_revision_of_file(@repo,
+                                              @repo/@file),
+      ).to eq("blah")
+    end
+
+    it "returns last revision of file based on before_commit" do
+      expect(
+        described_class.last_revision_of_file(@repo, @repo/@file,
+                                              before_commit: "0..3"),
+      ).to eq("brew")
+    end
+  end
+end
+
+describe Utils do
+  before(:each) do
+    if described_class.instance_variable_defined?(:@git)
+      described_class.send(:remove_instance_variable, :@git)
+    end
+  end
+
+  describe "::git_available?" do
+    it "returns true if git --version command succeeds" do
+      expect(described_class.git_available?).to be_truthy
+    end
+
+    it "returns false if git --version command does not succeed" do
+      stub_const("HOMEBREW_SHIMS_PATH", mktmpdir/"shim")
+      expect(described_class.git_available?).to be_falsey
+    end
+  end
+
+  describe "::git_path" do
+    it "returns nil when git is not available" do
+      stub_const("HOMEBREW_SHIMS_PATH", mktmpdir/"shim")
+      expect(described_class.git_path).to eq(nil)
+    end
+
+    it "returns path of git when git is available" do
+      expect(described_class.git_path).to end_with("git")
+    end
+  end
+
+  describe "::git_version" do
+    it "returns nil when git is not available" do
+      stub_const("HOMEBREW_SHIMS_PATH", mktmpdir/"shim")
```

Comment:
Instead of `mktmpdir` write to an already existing location that's deleted after tests: for example: https://github.com/Homebrew/brew/blob/2c43e069c25e3c7cc255f5cd105df3597246b952/Library/Homebrew/test/spec_helper.rb#L110-L129